resqpy/olio/base.py

```diff
@@ -0,0 +1,163 @@
+"""Base class for generic resqml objects """
+
+import logging
+import warnings
+from abc import ABCMeta, abstractmethod
+
+import resqpy.olio.uuid as bu
+import resqpy.olio.xml_et as rqet
+
+
+logger = logging.getLogger(__name__)
+
+
+class BaseResqml(metaclass=ABCMeta):
+    """Base class for generic RESQML objects
+    
+    Implements generic attributes such as uuid, root, part, title, originator.
+
+    Implements generic magic methods, such as pretty printing and testing for
+    equality.
+    
+    Example use::
+
+        class AnotherResqmlObject(BaseResqml):
+            
+            _content_type = 'obj_anotherresqmlobjectrepresentation'
+
+    """
+
+    @property
+    @abstractmethod
+    def _content_type(self):
+        """Definition of which RESQML object the class represents.
+        
+        Subclasses must overwrite this abstract attribute.
+        """
+        raise NotImplementedError
+
+    def __init__(self, model, uuid=None, title=None, originator=None):
+        """Load an existing resqml object, or create new.
+
+        Args:
+            model (resqpy.model.Model): Parent model
+            uuid (str, optional): Load from existing uuid (if given), else create new.
+            title (str, optional): Citation title
+            originator (str, optional): Creator of object. By default, uses user id.
+        """
+        self.model = model
+        self.title = title
+        self.originator = originator
+
+        if uuid is None:
+            self.uuid = bu.new_uuid()
+            logger.debug(f"Created new uuid for object {self}")
+        else:
+            self.uuid = uuid
+            logger.debug(f"Loading existing object {self}")
+            self.load_from_xml()
+    
+    # Define attributes self.part and self.root, using uuid as the primary key
+
+    @property
+    def part(self):
+        """Part corresponding to self.uuid"""
+
+        if self.uuid is None:
+            raise ValueError('Cannot get part if uuid is None')
+        return self.model.part_for_uuid(self.uuid)
+
+    @property
+    def root(self):
+        """Node corresponding to self.uuid"""
+
+        if self.uuid is None:
+            raise ValueError('Cannot get root if uuid is None')
+        return self.model.root_for_uuid(self.uuid)
+    
+    @root.setter
+    def root(self, value):
+        """Update self.uuid to match new root. Ensure root is added as a part"""
+
+        new_uuid = rqet.uuid_for_part_root(value)
+        if new_uuid is None:
+            raise ValueError("Cannot set uuid to be None")
+        self.uuid = new_uuid
+
+    def load_from_xml(self):
```

Comment:
docstring could perhaps contain an explanation that this method only loads the common elements of the object, and that derived classes should complete the job.

Reply:
I was hoping we could extend this class further ( in #44 ) so that in the majority of cases the base class would handle basic XML and HDF attributes, so that many simple derived classes would not need to implement anything to complete the job.